### PR TITLE
Pulsar binder provisioner refactoring

### PR DESCRIPTION
**Pull request by sobychacko**

### CODE REVIEW

File: spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/config/PulsarBinderConfiguration.java [hunk 1]
```diff
@@ -22,21 +22,34 @@
...
+ */
 @Configuration(proxyBeanMethods = false)
 @ConditionalOnMissingBean(Binder.class)
 @EnableConfigurationProperties({ PulsarProperties.class, PulsarExtendedBindingProperties.class })
 public class PulsarBinderConfiguration {
 
 	@Bean
-	public PulsarTopicProvisioner pulsarTopicProvisioner() {
-		return new PulsarTopicProvisioner();
+	public PulsarBinderConfigurationProperties pulsarBinderConfigurationProperties() {
```
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/config/PulsarBinderConfiguration.java:45]: Can we move this to `@ECP` above like the other CP or is there a particular reason this is a `@Bean`?

File: spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/properties/PulsarBinderConfigurationProperties.java [hunk 2]
```diff
@@ -18,7 +18,22 @@
...
 
+/**
+ * {@link ConfigurationProperties} for Pulsar binder configuration.
+ *
+ * @author Soby Chacko
+ */
 @ConfigurationProperties(prefix = "spring.cloud.stream.pulsar.binder")
 public class PulsarBinderConfigurationProperties {
 
+	private int partitionCount = 1;
```
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/properties/PulsarBinderConfigurationProperties.java:29]: Do you mind adding a comment here (you may be waiting to do all the doc stuff at once later though).

File: spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java [hunk 3]
```diff
@@ -34,74 +36,63 @@
...
+	}
+
 	@Override
 	public ProducerDestination provisionProducerDestination(String name,
 			ExtendedProducerProperties<PulsarProducerProperties> properties) throws ProvisioningException {
-		PulsarTopic pulsarTopic = PulsarTopic.builder(name).numberOfPartitions(1).build();
-		return new PulsarProducerDestination(pulsarTopic.topicName(), pulsarTopic.numberOfPartitions());
+
+		int partitionCount = this.pulsarBinderConfigurationProperties.partitionCount();
+		int partitionCountOnBinding = properties.getPartitionCount(); // leverage the
```
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java:54]: The comment here confused me for a sec as "core" is overloaded. I know what you mean now (SCSt core) but I think the variable name says it all. I would be in favor of dropping the comment.

File: spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java [hunk 4]
```diff
@@ -34,74 +36,63 @@
...
+
+	public PulsarTopicProvisioner(PulsarAdministration pulsarAdministration,
+			PulsarBinderConfigurationProperties pulsarBinderConfigurationProperties) {
+		this.pulsarAdministration = pulsarAdministration;
+		this.pulsarBinderConfigurationProperties = pulsarBinderConfigurationProperties;
+	}
+
 	@Override
 	public ProducerDestination provisionProducerDestination(String name,
 			ExtendedProducerProperties<PulsarProducerProperties> properties) throws ProvisioningException {
```
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java:51]: Wdyt about naming `properties -> bindingProps`?

File: spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java [hunk 5]
```diff
@@ -34,74 +36,63 @@
...
-		PulsarTopic pulsarTopic = PulsarTopic.builder(name).numberOfPartitions(1).build();
-		return new PulsarProducerDestination(pulsarTopic.topicName(), pulsarTopic.numberOfPartitions());
+
+		int partitionCount = this.pulsarBinderConfigurationProperties.partitionCount();
+		int partitionCountOnBinding = properties.getPartitionCount(); // leverage the
+																		// partition count
+																		// from core
+		if (partitionCountOnBinding > 1) {
```
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java:57]: Maybe we can make the default for the binding props PC `Integer null` and then validate when set > 0. 

We are using the default `1` here to signal that the user has set a value.

File: spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java [hunk 6]
```diff
@@ -34,74 +36,63 @@
...
 
 	@Override
 	public ConsumerDestination provisionConsumerDestination(String name, String group,
 			ExtendedConsumerProperties<PulsarConsumerProperties> properties) throws ProvisioningException {
-		PulsarTopic pulsarTopic = PulsarTopic.builder(name).numberOfPartitions(1).build();
-		return new PulsarConsumerDestination(pulsarTopic.topicName(), pulsarTopic.numberOfPartitions());
+		int partitionCount = this.pulsarBinderConfigurationProperties.partitionCount();
+
+		int partitionCountOnBinding = properties.getExtension().getPartitionCount(); // partition
+																						// count
```
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java:71]: **[NIT]** The formatter does this on the end of line comments. Do you mind moving it up on the line above the code? 

One of the most confusing areas of SCSt (for me anyways) is the extended props. I would love to see a comment like this: 

```
// The spring.cloud.stream.pulsar.bindings.<NAME>.consumer.partition-count 
```

File: spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java (hunk 5 above)
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java:57]: We could factor out the common code of these 2 and just pass in the 2 properties to the function.

File: spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java [hunk 7]
```diff
@@ -34,74 +36,63 @@
...
+																						// binding
+		if (partitionCountOnBinding > 1) {
+			partitionCount = partitionCountOnBinding;
+		}
+		PulsarTopic pulsarTopic = PulsarTopic.builder(name).numberOfPartitions(partitionCount).build();
+		this.pulsarAdministration.createOrModifyTopics(pulsarTopic);
+		return new PulsarDestination(pulsarTopic.topicName(), pulsarTopic.numberOfPartitions());
 	}
 
-	private static class PulsarDestination implements ProducerDestination, ConsumerDestination {
```
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/provisioning/PulsarTopicProvisioner.java:52]: Nice! `record` FTW!

File: spring-pulsar-spring-cloud-stream-binder/src/test/java/org/springframework/pulsar/spring/cloud/stream/binder/PulsarBinderUtilsTests.java [hunk 8]
```diff
@@ -25,6 +25,9 @@
 import org.springframework.cloud.stream.provisioning.ConsumerDestination;
 import org.springframework.pulsar.spring.cloud.stream.binder.properties.PulsarConsumerProperties;
 
+/**
+ * @author Soby Chacko
```
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/test/java/org/springframework/pulsar/spring/cloud/stream/binder/PulsarBinderUtilsTests.java:29]: **[NIT]** Do you mind adding the boilerplate ` Unit tests for {@link PulsarBinderUtils}.` description?

File: spring-pulsar-spring-cloud-stream-binder/src/test/java/org/springframework/pulsar/spring/cloud/stream/binder/PulsarTopicProvisionerTests.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+				pulsarBinderConfigurationProperties);
+		ExtendedProducerProperties<PulsarProducerProperties> properties = new ExtendedProducerProperties<>(
+				new PulsarProducerProperties());
+		ProducerDestination producerDestination = pulsarTopicProvisioner.provisionProducerDestination("foo",
+				properties);
+		verifyAndAssert(pulsarAdministration, producerDestination.getName(), "foo", 1);
+	}
+
+	private static void verifyAndAssert(PulsarAdministration pulsarAdministration, String producerDestination,
+			String foo, int expected) {
```
Comment (onobc) [inline spring-pulsar-spring-cloud-stream-binder/src/test/java/org/springframework/pulsar/spring/cloud/stream/binder/PulsarTopicProvisionerTests.java:57]: I think renaming the args in the `actual / expected` pattern makes this easier to follow:
```java
producerDestination -> actualProducerDestination
foo -> expectedProducerDestination
expect -> expectedPartitionCount
```
wdyt?

**Review by onobc — APPROVED** [review]
@sobychacko looks good. My comments are pretty much nits about comments, naming etc.. Feel free to update what you agree w/ and then merge.

I'm not sure whats up w/ the CI though. Please take a look before merging.

File: spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/properties/PulsarBinderConfigurationProperties.java (hunk 2 above)
Comment (sobychacko) [reply spring-pulsar-spring-cloud-stream-binder/src/main/java/org/springframework/pulsar/spring/cloud/stream/binder/properties/PulsarBinderConfigurationProperties.java:29]: i will address the javadocs in one go later.

File: spring-pulsar-spring-cloud-stream-binder/src/test/java/org/springframework/pulsar/spring/cloud/stream/binder/PulsarTopicProvisionerTests.java (hunk 9 above)
Comment (sobychacko) [reply spring-pulsar-spring-cloud-stream-binder/src/test/java/org/springframework/pulsar/spring/cloud/stream/binder/PulsarTopicProvisionerTests.java:57]: Good idea.

**Pull request merged**